tlv/record_type.go

```diff
@@ -0,0 +1,85 @@
+package tlv
+
+import (
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/lightningnetwork/lnd/fn"
+	"golang.org/x/exp/constraints"
+)
+
+// RecordT is a high-order type makes it easy to encode known "primitive" types
+// as TLV records.
+type RecordT[T TlvType, V any] struct {
+	// recordType is the type of the TLV record.
+	recordType T
+
+	// val is the value of the underlying record. Go doesn't let us just
+	// embed the type param as a struct field, so we need an intermediate
+	// variable.
+	Val V
+}
+
+// RecordProducerT is a type-aware wrapper around the normal RecordProducer
+// interface.
+type RecordProducerT[T any] interface {
+	RecordProducer
+
+	// This is a non-interface type constraint that allows us to pass a
+	// concrete type as a type parameter rather than a pointer to the type
+	// that satisfies the Record interface.
+	*T
+}
+
+// NewRecordT creates a new RecordT type from a given RecordProducer type. This
+// is useful to wrap a given record in this utility type, which also serves as
+// an extra type annotation. The underlying type of the record is retained.
+func NewRecordT[T TlvType, K any, V RecordProducerT[K]](
```

Comment:
Why not `K RecordProducer` instead of `K any, V RecordProducerT[K]`

Reply:
If it's just `K RecordProducer`, then you can't do `RecordT[TlvType, wireCsv]`. You need to do `RecordT[TlvType, *wireCsv]` since `RecordProducer` needs a pointer receiver for the method. 

The trick I'm using is described here: https://go.googlesource.com/proposal/+/refs/heads/master/design/43651-type-parameters.md#pointer-method-example

We're basically saying that `K` is anything, but it should be able to be passed anywhere a `RecordProducerT` can be passed in, and that asserts that `K` has a pointer method to satisfy that interface. If you play with it locally, ensuring the tests compile, this is the only way you can get it to work. 

The prior version of just `RecordT[TlvType, *wireCsv]`  worked, but then if you had some struct `wireCsv`, you couldn't just do `var msg wireCsv`, and decode into that. You'd need to instantiate all internal variables first, since it's actually a pointer. This lets us pass in a type `T`, but make sure that the `*T` version of it can satisfy an interface we need.

Reply:
You still get the type safety tho, if I try to pass in something that doesn't implement `RecordProducer`, you get an error like: 
```
./record_type_test.go:74:33: *wireCsv does not satisfy RecordProducerT[wireCsv] (missing method Record)
```